pyEpiabm/pyEpiabm/routine/file_population_config.py

```diff
@@ -0,0 +1,195 @@
+#
+# Factory for creation of a population based on an input file
+#
+
+import numpy as np
+import pandas as pd
+import random
+import copy
+
+from pyEpiabm.core import Household, Population, Cell
+from pyEpiabm.core.microcell import Microcell
+from pyEpiabm.property import InfectionStatus
+
+
+class FilePopulationFactory:
+    """ Class that creates a population  based on an input .csv file.
+    """
+    @staticmethod
+    def make_pop(input_file: str, random_seed: int = None):
+        """Initialize a population object from an input csv file, with one
+        row per microcell. A uniform multinomial distribution is
+        used to distribute the number of people into the different households
+        within each microcell. A random seed may be specified for reproducible
+        populations.
+
+        Input file contains columns:
+            * `cell`: ID code for cell (Uses hash if unspecified)
+            * `microcell`: ID code for microcell (Uses hash if unspecified)
+            * `location_x`: The x coordinate of the parent cell location
+            * `location_y`: The y coordinate of the parent cell location
+            * `household_number`: Number of households in that microcell
+            * Any number of columns with titles from the `InfectionStatus`
+                enum (such as `InfectionStatus.Susceptible`), giving the
+                number of people with that status in that cell
+
+        :param input_file: Path to input file
+        :type input_file: str
+        :param random_seed: Seed for reproducible household distribution
+        :type random_seed: int
+        :return: Population object with individuals distributed into
+            households
+        :rtype: Population
+        """
+        # If random seed is specified in parameters, set this
+        if random_seed is not None:
+            np.random.seed(random_seed)
+            random.seed(random_seed)
+
+        # Read file into pandas dataframe
+        input = pd.read_csv(input_file)
+        loc_given = ("location_x" and "location_y" in input.columns.values)
+
+        # Validate all column names in input
+        valid_names = ["cell", "microcell", "location_x",
+                       "location_y", "household_number"]
+        for col in input.columns.values:  # Check all column headings
+            if not (hasattr(InfectionStatus, col) | (col in valid_names)):
+                raise ValueError(f"Unknown column heading '{col}' in input")
```

Comment:
To be honest, this was intended to be helpful - say a user has a column "InfctMild" then this will highlight the typo as opposed to letting it pass silently and then giving unusual simulation outputs.

While I recognise this is inconvenient to users who want to import their data from other sources and so have other columns that are not used here, I am inclined to leave this as it is simply because I hope to have added a logging future in the short/mid term future, which will make this much less unpleasant as extra columns would simply be listed as a warning in there, which could be ignored if it was intentional.

I have not explicitly tested that all required columns are present as the default error messages are pretty informative here anyway (and this is not something we would want to handle later in logging - if there is no cell ID the population literally can't be built meaningfully!)

Reply:
Sure, your reasoning makes sense, thanks.